lua/packer/plugin_types/git.lua

```diff
@@ -264,12 +264,22 @@ git.setup = function(plugin)
                                   {success_test = exit_ok, capture_output = update_callbacks}))
         :map_err(function(err)
           plugin.output = {err = vim.list_extend(update_info.err, update_info.output), data = {}}
-
-          return {
-            msg = fmt('Error pulling updates for %s: %s', plugin_name,
-                      table.concat(update_info.output, '\n')),
-            data = err
-          }
+          -- TODO should retrying be inside of the map error function
+          local git_err = update_info.err and update_info.err[1] or nil
+          if git_err and git_err:match(vim.pesc("Not possible to fast-forward")) then
+            local update_with_rebase = update_cmd:gsub('rebase=false', 'rebase=true')
+            r = await(jobs.run, update_with_rebase,
```

Comment:
@wbthomason so the advice I needed was how to use the current job/async module to handle hitting an error and wanting to start off another job. Can I just do this inside of the `map_err` as I've done. This sort of works, as in the logic is triggered (though incomplete) but looks kind of wrong (for no concrete reasons, just a feeling).

Or is there some other mechanism for this or should there be?

Reply:
The `Result` type semantics are uglier than I wanted them to be in general; I've thought about redesigning that mechanism but haven't had any great ideas for it yet.

The better way to do what you want here is with `or_else`: this takes and runs a result-returning function, so you could pass it this function doing the retry and return the result of your `await():and_then` sequence.

For what it's worth, the `Result` "type" is meant to provide similar semantics to e.g. https://doc.rust-lang.org/std/result/ or Haskell's `Either`, if you're familiar with those. Unfortunately, Lua's syntax (e.g. repeating `function()` everywhere) make it rather ugly.

Reply:
Ah @wbthomason that makes sense I wasn't aware of the `or_else` method. Re. `Result` and `Eithers` and other monad/monad like patterns I'm generally a big fan of. Will give that a go 👍🏾